client/state/happychat/reducer.js

```diff
@@ -0,0 +1,90 @@
+import { combineReducers } from 'redux';
+import get from 'lodash/get';
+import find from 'lodash/find';
+import concat from 'lodash/concat';
+
+import {
+	HAPPYCHAT_SET_MESSAGE,
+	HAPPYCHAT_RECEIVE_EVENT,
+	HAPPYCHAT_CONNECTING,
+	HAPPYCHAT_CONNECTED,
+} from 'state/action-types';
+
+/**
+ * Returns a timeline event from the redux action
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ *
+ */
+const timeline_event = ( state = {}, action ) => {
+	switch ( action.type ) {
+		case HAPPYCHAT_RECEIVE_EVENT:
+			const event = action.event;
+			return Object.assign( {}, {
+				id: event.id,
+				message: event.message,
+				nick: event.user.nick,
+				image: event.user.picture,
+				user_id: event.user.id,
+				type: event.type,
+				links: get( event, 'meta.links' )
+			} );
+	}
+	return state;
+};
+
+/**
+ * Adds timeline events for happychat
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ *
+ */
+const timeline = ( state = [], action ) => {
+	switch ( action.type ) {
+		case HAPPYCHAT_RECEIVE_EVENT:
+			const event = timeline_event( {}, action );
+			const existing = find( state, ( { id } ) => event.id === id );
+			return existing ? state : concat( state, [ event ] );
```

Comment:
Not an issue right now, but in future if we ever want to allow edits to messages, this would be a problem since we skip the event if it already exists.